### PR TITLE
Allow loading projects as either a single tarball (default) or expanded files

### PR DESCRIPTION
**Pull request by humphd**
This alters the Thimble client and server to allow for a project to be loaded either as a tarball, or as a set of separate files.  I'm doing this behind an opt-in flag, because I want to experiment with it.  I'm curious to see what will happen if we remove the tarball creation from publish.  It is not fully optimized (e.g., I would need to modify publish to send `Buffer` vs. JSON), but that shouldn't matter too much.

To make this work, I've changed the project load order to load the metadata first.  This gives the file IDs if I need them for loading an expanded set of files.

To use it, add `?loadfiles=1` to the Thimble URL.  If you do nothing, it should load the project as a tarball like normal.

### CODE REVIEW

File: public/editor/scripts/project/metadata.js [hunk 1]
```diff
@@ -279,24 +284,12 @@ define(function(require) {
...
-          data: data,
-          id: config.id,
-          root: config.root,
-          user: config.user,
-          title: config.title
-        }, callback);
-      });
+  function install(config, callback) {
+    if(!config.user) {
+      loadAnonymous(config, callback);
```
Comment (humphd) [inline public/editor/scripts/project/metadata.js]: um, yup!  Good catch.

**Comment by humphd** [comment]
I noticed that error, too, and filed it at https://github.com/mozilla/thimble.mozilla.org/issues/2257.  I think it might be that missing `return` you found.

I've fixed everything you noticed, want to test again?

**Comment by gideonthomas** [comment]
I still see the failure with the `upgrade` workflow I mention above.

**Comment by humphd** [comment]
OK, thanks, I'll debug it locally.  I was hoping that would be all it was.

**Comment by humphd** [comment]
I tried to fix this quickly tonight, but I think I'm going to have to spend a bunch of time on it to get it right.  The problem is that we used to create the project dir when we grabbed the tarball, but now I need it sooner so I can start writing file attributes.  I'll try and do that in the next few days.

**Comment by humphd** [comment]
@gideonthomas I think this is working well now, can you re-test please?  You can change your `env` for the load type to change it from `"tarball"` to `"files"`.

File: public/editor/scripts/project/metadata.js [hunk 2]
```diff
@@ -226,7 +226,7 @@ define(function(require) {
       // [{ id: 1, path: "/index.html", project_id: 3 }, ... ]
       if (config.data) {
         project.id = config.id;
-
+console.log("config.data", config.data);
```
Comment (gideonthomas) [inline public/editor/scripts/project/metadata.js]: remove

File: server/routes/main/editor.js [hunk 3]
```diff
@@ -6,9 +6,13 @@ var querystring = require("querystring");
 var defaultProjectNameKey = require("../../../constants").DEFAULT_PROJECT_NAME_KEY;
 var utils = require("../utils");
 var HttpError = require("../../lib/http-error");
-
+var env = require("../../lib/environment");
 var snippets = require("../../lib/snippets");
 
+// One of "tarball" or "files" to specify how projects should get loaded.
+// If not specified, the client code will default to "tarball".
+var projectLoadStrategy = env.get("PROJECT_LOAD_STRATEGY");
```
Comment (gideonthomas) [inline server/routes/main/editor.js]: can we do this in server/routes/config.js instead and pass in `config.projectLoadStrategy` below?

**Comment by gideonthomas** [comment]
r+

**Pull request merged**